openedx/core/djangoapps/schedules/tasks.py

```diff
@@ -153,23 +160,203 @@ def absolute_url(relative_path):
+        target_date=beginning_of_day,
+        bin_num=bin_num,
+        num_bins=RECURRING_NUDGE_NUM_BINS,
+    )
+
+    schedules = get_schedules_with_target_date_by_users_and_orgs(
+        schedule_date_field='upgrade_deadline',
+        target_date=beginning_of_day,
+        users=users,
+        org_list=org_list,
+        exclude_orgs=exclude_orgs,
+    )
+
+    LOG.debug('Upgrade Reminder: Query = %r', schedules.query.sql_with_params())
+
+    for schedule in schedules:
+        enrollment = schedule.enrollment
+        user = enrollment.user
+
+        course_id_str = str(enrollment.course_id)
+
+        # TODO: group by schedule and user like recurring nudge
+        course_id_strs = [course_id_str]
+        first_schedule = schedule
+
+        template_context = get_base_template_context()
+        template_context.update({
+            'student_name': user.profile.name,
+            'user_personal_address': user.profile.name if user.profile.name else user.username,
+            'user_schedule_upgrade_deadline_time': dateformat.format(
+                schedule.upgrade_deadline,
+                get_format(
+                    'DATE_FORMAT',
+                    lang=first_schedule.enrollment.course.language,
+                    use_l10n=True
+                )
+            ),
+
+            'course_name': first_schedule.enrollment.course.display_name,
+            'course_url': absolute_url(reverse('course_root', args=[str(first_schedule.enrollment.course_id)])),
+
+            # This is used by the bulk email optout policy
+            'course_ids': course_id_strs,
+        })
+
+        yield (user, first_schedule.enrollment.course.language, template_context)
+
+
+def get_users_with_target_date_by_bin(schedule_date_field, target_date, bin_num, num_bins=DEFAULT_NUM_BINS):
+    """Returns queryset of Users that have schedules that match the target_date and whose id matches the bin_num.
+
+    Arguments:
+    schedule_date_field -- string field name to query on the User's Schedule model
+    target_date -- datetime day (with zeroed-out time) that the User's Schedule's schedule_date_field value should fall
+                   under
+    bin_num -- int for selecting the bin of Users whose id % num_bins == bin_num
+    num_bin -- int specifying the number of bins to separate the Users into (default: DEFAULT_NUM_BINS)
+    """
+    schedule_date_equals_target_date_filter = {
+        'courseenrollment__schedule__{}__gte'.format(schedule_date_field): target_date,
+        'courseenrollment__schedule__{}__lt'.format(schedule_date_field): target_date + datetime.timedelta(days=1),
+    }
+    return User.objects.filter(
+        courseenrollment__is_active=True,
+        **schedule_date_equals_target_date_filter
+    ).annotate(
+        id_mod=F('id') % num_bins
+    ).filter(
+        id_mod=bin_num
+    )
+
 
+def get_schedules_with_target_date_by_users_and_orgs(schedule_date_field, target_date, users, org_list=None,
```

Comment:
I was envisioning one method that did both the user and schedule stuff. Do we need to separate it into two?